hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/keyvalue/KeyValueHandler.java

```diff
@@ -1483,21 +1489,249 @@ public void deleteContainer(Container container, boolean force)
+    }
+
+    BlockData putBlockData = BlockData.getFromProtoBuf(peerBlockData);
+    putBlockData.setChunks(successfullChunksList);
+    putBlockForClosedContainer(container, putBlockData, maxBcsId, overwriteBcsId);
+  }
+
+  private ByteString readChunkData(XceiverClientSpi xceiverClient, ContainerProtos.ChunkInfo chunkInfoProto,
+                                   BlockID blockID, Token<OzoneBlockTokenIdentifier> blockToken) throws IOException {
+    ContainerProtos.ReadChunkResponseProto response =
+        ContainerProtocolCalls.readChunk(xceiverClient, chunkInfoProto, blockID.getDatanodeBlockIDProtobuf(),
+            null, blockToken);
+
+    if (response.hasData()) {
+      return response.getData();
+    } else if (response.hasDataBuffers()) {
+      return BufferUtils.concatByteStrings(response.getDataBuffers().getBuffersList());
+    } else {
+      throw new IOException("Error reading chunk data: No data returned.");
+    }
+  }
+
+  /**
+   * This method reconciles chunks per block. It reads the missing/corrupt chunk data from the peer
+   * datanode and writes it to the local container. If the chunk write fails, the block commit sequence
+   * id is not updated.
+   */
+  private void reconcileChunksPerBlock(KeyValueContainer container, ContainerData containerData,
+                                       TokenHelper tokenHelper, XceiverClientSpi xceiverClient, long blockId,
+                                       List<ContainerProtos.ChunkMerkleTree> chunkList) throws IOException {
+    Set<Long> offsets = chunkList.stream().map(ContainerProtos.ChunkMerkleTree::getOffset)
+        .collect(Collectors.toSet());
+    BlockID blockID = new BlockID(containerData.getContainerID(), blockId);
+    // The length of the block is not known, so instead of passing the default block length we pass 0. As the length
+    // is not used to validate the token for getBlock call.
+    Token<OzoneBlockTokenIdentifier> blockToken = tokenHelper.getBlockToken(blockID, 0L);
+    ContainerProtos.GetBlockResponseProto blockResponse = ContainerProtocolCalls.getBlock(xceiverClient, blockID,
+        blockToken, new HashMap<>());
+    ContainerProtos.BlockData peerBlockData = blockResponse.getBlockData();
+    BlockData localBlockData = getBlockManager().getBlock(container, blockID);
+    // Check the local bcsId with the one from the bcsId from the peer datanode.
+    long maxBcsId = Math.max(peerBlockData.getBlockID().getBlockCommitSequenceId(),
+            localBlockData.getBlockCommitSequenceId());
+    List<ContainerProtos.ChunkInfo> chunksListFromPeer = peerBlockData.getChunksList();
+
+    SortedMap<Long, ContainerProtos.ChunkInfo> localChunksMap = localBlockData.getChunks().stream()
+            .collect(Collectors.toMap(ContainerProtos.ChunkInfo::getOffset,
+                Function.identity(), (chunk1, chunk2) -> chunk1, TreeMap::new));
+    boolean overwriteBcsId = true;
+
+    for (ContainerProtos.ChunkInfo chunkInfoProto : chunksListFromPeer) {
```

Comment:
This negates the benefit of the merkle tree. The loop is `O(len(chunksListFromPeer))`. We want this iteration to be `O(len(chunkList))`, where `chunkList` is the parameter defined in this method as only the chunks that we are missing.

Reply:
I agree but if we had the chunkInfos in the `chunkList` that might be possible, Our only source of truth is `chunksListFromPeer` which has the chunkInfo along with checksum data. So even if we created a temp `chunkInfo` to just read the chunk. We need to validate the read data with chunk checksum, So we'd have to iterate through the `chunksListFromPeer` anyway.

Reply:
This has been fixed in the latest update.